Stash apply

ethomson opened the pull request:
A modification on #2705.  I made some notes in that PR, though actually acting on those notes was quite a bit more challenging than I had hoped.  This introduces two lower level ideas:

* `git_merge__iterators`, to allow one to merge a tree and an index and get an index out as a result.  (Interestingly `git_merge` had this broken out as a helper function at one point, and it was brought back in.  So now it's a helper function again.)  This allows us to merge the results of another merge, without having to write it out to a tree.

* `git_index_read_index`, which behaves like `git_index_read_tree` except that it reads an index into the index.  This maintains the stat cache of the existing items.

Sorry about the long delay in finishing this up.  Stash application is completely and totally ridiculous (put these files on disk and these in the index _unless_ this happens or this), as I'm sure @swisspol can attest to.  I think that with the two basic functions mentioned above that the `stash_apply` can be a bit simpler and not necessarily have to write objects for the intermediate steps.

swisspol commented:
Thanks @ethomson, awesome work :+1:

swisspol commented:
> This allows us to merge the results of another merge, without having to write it out to a tree.

BTW is this for performance reasons or API impedance reasons?

swisspol commented:
@ethomson Any idea when this will land?

ethomson commented:
Thank you @swisspol for all your awesome work - I literally cannot express how painful it was to understand how stash application actually works in Git.  Crazy.  So thanks for digging into this.

And again, sorry to bikeshed all over this (and to take so long to do so) - this was mostly for perf (I work in a very large tree, a write/read of that is expensive.)

Therzok commented:
> If any untracked or ignored file saved in the stash already exist in the
> workdir, the function will return GIT_EEXISTS and both the workdir and index
> will be left untouched.

This doesn't seem to return that, testing it under libgit2sharp as we speak.

swisspol commented:
Do you mean it returns a different error code or it succeeds?

Therzok commented:
It just returns GIT_EMERGECONFLICT. The tests that fail what the documentation states:

[Test 1](https://github.com/libgit2/libgit2sharp/pull/871/files#diff-0287b4e63ba41ee4a3a63320fdaecc6eR253) - It flushes the contents to the working directory regardless.

{Test 2](https://github.com/libgit2/libgit2sharp/pull/871/files#diff-0287b4e63ba41ee4a3a63320fdaecc6eR275) - Fails because it's not returning GIT_EEXISTS, but GIT_EMERGECONFLICT.

swisspol commented:
Any chance you could roll back libgit2 to c31a1ed and check if that behavior was in my original PR or a fallback of @ethomson changes?

Therzok commented:
I get different errors by following the docs with c31a1ed:

* With no REINSTATE_INDEX set, I get the index reinstated on
```
touch file
put text
add
save
pop
```

* With REINSTATE_INDEX not set, I don't get the index reinstated on:
```
touch file
put text
add
save
pop
```

* With REINSTATE_INDEX set, I was expecting to get conflicts on
```
touch file1, file2
put text in both
add file 1
stash
touch file1, file2
put other text in both
add file1
stash pop
```
But I got the stash applied successfully?

Probably _I_ did something wrong, someone should look at the tests to see if I approached them correctly.

swisspol commented:
> With REINSTATE_INDEX not set, I get the index reinstated on

This is the expected git stash behavior (don't ask, git stash is a mess):

```sh
polbookpro:test pol$ git init
Initialized empty Git repository in /private/tmp/test/.git/
polbookpro:test pol$ git commit --allow-empty -m "Initial commit"
[master (root-commit) 1ec5ac0] Initial commit
polbookpro:test pol$ date > file
polbookpro:test pol$ git add file
polbookpro:test pol$ git stash
Saved working directory and index state WIP on master: 1ec5ac0 Initial commit
HEAD is now at 1ec5ac0 Initial commit

polbookpro:test pol$ git stash pop
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	new file:   file

Dropped refs/stash@{0} (341999004335d04dbc782d2dce198b2ec3e56d03)
```

And again with the `--index` option:
```
polbookpro:test pol$ git stash
Saved working directory and index state WIP on master: 1ec5ac0 Initial commit
HEAD is now at 1ec5ac0 Initial commit
polbookpro:test pol$ git stash pop --index
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	new file:   file

Dropped refs/stash@{0} (d1c4bc6c9be9d331285b32d22876663aa69990f5)
```

Same end result, I know makes no sense. That's why I wrote stash API in libgit2 to mirror exactly how it's implemented in Git CLT (even if it was so convoluted for no good reason). You have all these peculiar behaviors you need to preserve.

> With REINSTATE_INDEX set, I was expecting to get conflicts on

Yeah this one should fail
```sh
polbookpro:test pol$ git init
Initialized empty Git repository in /private/tmp/test/.git/
polbookpro:test pol$ git commit --allow-empty -m "Initial commit"
[master (root-commit) 41e9e67] Initial commit
polbookpro:test pol$ date > file1
polbookpro:test pol$ date > file2
polbookpro:test pol$ git add file1
polbookpro:test pol$ git stash
Saved working directory and index state WIP on master: 41e9e67 Initial commit
HEAD is now at 41e9e67 Initial commit
polbookpro:test pol$ date > file1
polbookpro:test pol$ date > file2
polbookpro:test pol$ git add file1
polbookpro:test pol$ git stash pop
Auto-merging file1
CONFLICT (add/add): Merge conflict in file1
polbookpro:test pol$ git status
On branch master
Unmerged paths:
  (use "git reset HEAD <file>..." to unstage)
  (use "git add <file>..." to mark resolution)

	both added:      file1

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	file2

no changes added to commit (use "git add" and/or "git commit -a")
```

swisspol commented:
(I updated my previous comment, so read it on the web not in the email notification.)

Anyway as far as I can tell c31a1ed is working correctly and like Git CLT. However, it seems @ethomson's changes might have broken something.

Therzok commented:
Okay, I'm starting to understand how it works. Thanks for the clarifications.

I'll study git.git behavior on stash and update the tests to match that behavior. Thanks a lot.

swisspol commented:
@ethomson I just found this tiny bug, you should pick it up if you haven't already fixed it:
https://github.com/git-up/libgit2/commit/913662b202efb5e4ceac89766da96841bc7bbbb4

Therzok commented on src/stash.c:
```diff
@@ -553,6 +549,257 @@ int git_stash_save(
+	if ((flags & GIT_APPLY_REINSTATE_INDEX) &&
+		git_oid_cmp(git_tree_id(stash_parent_tree), git_tree_id(index_tree))) {
+
+		if ((error = merge_index_and_tree(
```
It seems that the [checkout options](https://github.com/libgit2/libgit2/blob/master/include/git2/checkout.h#L262) for notifying checkout paths isn't used here and below, right?

It's only reporting untracked files for me.

Therzok replied:
/cc @ethomson

ethomson replied:
> It seems that the checkout options for notifying checkout paths isn't used here and below, right?

> It's only reporting untracked files for me.

Can you provide a unit test or something to repro this with?

ethomson commented:
Yes, I dropped the `GIT_EEXISTS` since that's not quite what that error message means, I don't think.  I meant to create a new error type here, but forgot when I pushed this up.

Otherwise, I'm having a bit of trouble following this thread.  Can one of you two summarize the status here?  Is something else broken?  If so, what?

Therzok commented:
I was unaware of git stash's behavior in case of reinstate index or normal.

If EEXISTS was dropped, then it's fine by me.

Otherwise, there's no checkout notification called for the updated files.

ethomson commented:
So...  I restored the `GIT_EEXISTS` behavior when an untracked file exists in the working directory that would be overwritten by the unstash operation.  However I still feel like this is not quite the behavior that we want.

Do we really need two different exit codes here?  Could we use the checkout notify callbacks to discern between this case and other checkout conflicts?  I would much prefer to simply return `GIT_EMERGECONFLICT` here...

Otherwise, @Therzok - I think the behavior that you're seeing was because if you explicitly provide a checkout option, then you must also provide a checkout strategy that is not NONE (or else we won't do anything)...  I added a test here, but:

I added a change to imply a checkout strategy of `SAFE` when you use a checkout strategy (eg, from `GIT_CHECKOUT_STRATEGY_INIT`).  This better matches rebase, but it implies that you cannot have an unstash dry-run.  Meh.  I would like to think a little bit more about this because I think that we're trying to make it friendly for consumers (upgrading their checkout strategy) but cutting off the utility of `GIT_CHECKOUT_NONE`.)

swisspol commented:
> Otherwise, I'm having a bit of trouble following this thread. Can one of you two summarize the status here? Is something else broken? If so, what?

Sorry @ethomson I missed some comments here. The point is AFAIK my original implementation is "canonical" since based on exactly the same heuristics as Git shell script. @Therzok noticed some suspect behavior which partially went away when switching back to my original implementation. So looks one of your change introduced a regression. Not sure why it wasn't captured by the existing unit tests though.

Looks like with your latest changes in, @Therzok would need to confirm his tests behave same with both my original implementation and yours.

> I added a change to imply a checkout strategy of SAFE when you use a checkout strategy

It's like `git_reset()` too and seems reasonable to me.

ethomson commented:
My understanding is that @therzok was testing to the documentation, which I had not updated to reflect my removal of the `GIT_EEXISTS` behavior.

I have since updated to replace that behavior that I had removed but I think this is the wrong way to signal that condition and we should find something superior.

ethomson commented:
Okay, well, there was a resounding silence there.  So I've dropped the whole `GIT_EEXISTS` return code.  We now return `GIT_EMERGECONFLICT` for all merge conflicts, just like we do with all the other checkout flows.

I've added a progress callback to the options structure that you can use if you want to be able to discern one checkout conflict from another.

Seem sensible?

swisspol commented:
Where are we standing unit test wise now? Are the unit tests I originally written all pass *unmodified* with your latest version? That's the most important thing to me. Then I'm curious to hear about @therzok's functional tests and if the results are now the same for both my original implementation and yours.

I'm afraid I don't have feedback on the `GIT_EEXISTS` code as I'm not affected by it and I can't say I fully understand what's at stake, sorry.

swisspol commented:
I'm completely fine with the modified implementation BTW, I'd just like confirmation that the tests are passing all the way up the stack.

ethomson commented:
The only change that I made to your tests was to assert the return code to `GIT_EMERGECONFLICT`, then change it back to `GIT_EEXISTS`, and now back again.  Your tests were very helpful and comprehensive, thank you.

swisspol commented:
Sounds like that if @therzok's tests are now passing, this should most likely land finally :)

Therzok commented:
Yes, it's all working properly now. I'll update all of libgit2sharp's tests tomorrow.

I'm only curious as of why aren't the Checkout options themselves used for providing the notification callbacks. But this works for me.

ethomson commented:
Well, there's two checkouts...  so you may need to determine which one of them had the `GIT_EMERGECONFLICT`s...

I would love another set of eyes from @carlosmn though, if he has a moment.

Therzok commented:
I'm just wondering if you can't push updates to the checkout_options notifications, since most likely, a user will need to know _what_ happened, not _where from_. That's my $0.02 here.

ethomson commented:
If all you're concerned about is what happened then the checkout callbacks are appropriate and you should use them.  But if you care about whether those errors  were during the index restoration then checkout callbacks can't help you with that.

Aside from that, animating a progress bar is hard if you have to wait until the first checkout to see progress.

ethomson commented:
Gah, fat fingered the close button on my phone.

carlosmn commented:
The newly-added commits all seem sensible.

carlosmn commented:
This sounds like it should have its changes in the CHANGELOG; could we get a commit which adds the information there?

swisspol commented:
Hey @ethomson you wanna add this trivial fix to your PR since you're touching all this code?
```
diff --git a/src/stash.c b/src/stash.c
index dab3255..90411ca 100644
--- a/src/stash.c
+++ b/src/stash.c
@@ -111,7 +111,7 @@ static int build_tree_from_index(git_tree **out, git_index *index)
        git_oid i_tree_oid;
 
        if ((error = git_index_write_tree(&i_tree_oid, index)) < 0)
-               return -1;
+               return error;
 
        return git_tree_lookup(out, git_index_owner(index), &i_tree_oid);
 }
```

ethomson commented:
Cool.  I updated this with `CHANGELOG` information, and a little bit more documentation around the `checkout_options` change.

I also included @swisspol 's patch (above).

swisspol commented:
I was under the impression the unit tests I added to libgit2 were covering all cases, but I must have missed something as there's a regression with this PR compared to my original implementation (some of my higher-level functional tests now fail):

Start with this state:
```sh
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	new file:   bar.txt
	modified:   hello_world.txt

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   foo.txt

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	test.txt
```

Then do `git stash -u` and verify that `git status` shows no pending changes afterwards.

Using `git stash apply` results in this state:
```
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	new file:   bar.txt

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   foo.txt
	modified:   hello_world.txt

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	test.txt
```

However using libgit2's `git_stash_apply()` results in this different state:
```
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   foo.txt
	modified:   hello_world.txt

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	bar.txt
	test.txt

no changes added to commit (use "git add" and/or "git commit -a")
```

Notice that `bar.txt` wasn't restored in the index.

swisspol commented:
And here's another regression (possibly the same underlying issue):

Start with this state:
```
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	new file:   bar.txt
	modified:   hello_world.txt

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   foo.txt

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	test.txt
```

Then do git `stash -u` and verify that `git status` shows no pending changes afterwards.

Using `git stash apply --index` results in this state:
```
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	new file:   bar.txt
	modified:   hello_world.txt

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   foo.txt

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	test.txt
```

However using libgit2's `git_stash_apply()` with the `GIT_STASH_APPLY_REINSTATE_INDEX` flag results in this different state (which wasn't the case in my original implementation):
```
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   foo.txt
	modified:   hello_world.txt

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	bar.txt
	test.txt
```

ethomson commented:
I feel like there should be unit tests for this!  (In the libgit2 tests, I mean.)

swisspol commented:
I agree. What do you make of the 2nd case though? It seems `GIT_STASH_APPLY_REINSTATE_INDEX` is not working at all?

swisspol commented:
Hey @ethomson so what do you wanna do about the above issues? Current stash apply in libgit2 is not working like Git core unfortunately.

ethomson commented:
I think that the first step would be to add failing unit tests.  I'm heads-down right now, so if that's something you can do, that would be great.

swisspol commented:
Sorry for the noise above... It's because I have to continuously revert this stash apply implementation in my fork of libgit2, since it doesn't behave like Git CLT.

swisspol commented:
I created a dedicated issue to track this regression: #3230.
